---
title: "VCF Filtering Script"
author: "Nicholas F. Brazeau"
date: "`r format(Sys.time(), '%B %d, %Y')`"
output:
  prettydoc::html_pretty:
    theme: cayman
    highlight: vignette
    toc: true
editor_options: 
  chunk_output_type: console
---


```{r imports, include=F, warning=F, results="hide"}

# -----------------------------------
# The purpose of this script is to filter an example vcf 
# Filtering a VCF has many different approaches/theories (i.e. this is no "recipe" or "truly correct way") -- this is an example and is in no way comprehensive
# See https://gatkforums.broadinstitute.org/gatk/discussion/2806/howto-apply-hard-filters-to-a-call-set
# -----------------------------------


## This is where we import libraries and functions
library(vcfR)
library(tidyverse)
library(RColorBrewer)
library(grid)
library(DT)
devtools::install_github("nickbrazeau/NFBtools")
library(NFBtools)

```

```{r read_in_file, include=F, warning=F, results="hide"}
vcffile <- "<path to your vcffile>"
vcf <- vcfR::read.vcfR(vcffile)
```
# Introduction
# YOU NEED TO CHANGE THIS BASED ON YOUR PROJECT 
This is the VCF filter document _____ project. Thus far, I have done the following: 

1. Aligned with `bwa-mem` with default parameeters
2. Deduplicated with `Picard MarkDuplicates` and mate-fixed with `Picard FixMateInformation`
3. Ran `GATK RealignerIntervals` for local realignment (legacy not neccessary with HC really)
4. Variant called with `GATK HaplotypeCaller`
5. Built a variant annotation file with `snpEff` 


## Summary 
Overall, this VCF contains `r nrow(vcf@gt)` variants across `r (ncol(vcf@gt)-1)` samples.

The goal of this script is to filter this VCF following [GATK best practices hard filtering](http://gatkforums.broadinstitute.org/gatk/discussion/2806/howto-apply-hard-filters-to-a-call-set). Throughout this script we will  present summary statistics and plots for various `INFO` and `GT` field parameters that we plan on using for "Hard Filters" (see here for [VCF Specifications](https://samtools.github.io/hts-specs/VCFv4.2.pdf)).  


```{r subset_to_snps, include=F, warning=F, results="hide"}
vcf <-vcfR::extract.indels(vcf, return.indels = F) # subset to SNPs
vcf <- vcf[vcfR::is.biallelic(vcf)] # subset to biallelic SNPs

tidyvcf <- vcfR2tidy(vcf)
```
Overall, this VCF contains `r nrow(vcf@gt)` biallelic SNPs across `r (ncol(vcf@gt)-1)` samples.


### Summary Statistics for INFO Field Parameters
```{r extract_info_field_df, echo=F, include=T, warning=F, results="asis"}
infodf <- tidyvcf$fix

infodf_summary <- infodf %>% 
  select(c(ChromKey, QD, MQ, AF, DP, SOR, FS, MQRankSum, ReadPosRankSum)) %>% 
  dplyr::mutate_if(is.character, as.numeric) %>% 
  tidyr::gather(key="INFO_field_param", val="val", 2:ncol(.)) %>%
  dplyr::group_by(INFO_field_param) %>% 
  dplyr::summarise(Loci=n(),
                   Missing = sum(is.na(val))/Loci,
                   Min=min(val, na.rm=T),
                   Q10 = quantile(val, prob=0.1, na.rm=T),
                   Q25 = quantile(val, prob=0.25, na.rm=T),
                   Median = median(val, na.rm=T),
                   Mean = mean(val, na.rm=T),
                   Q75 = quantile(val, prob=0.75, na.rm=T),
                   Q90 = quantile(val, prob=0.9, na.rm=T),
                   Max = max(val, na.rm=T),
                   StDev = sd(val, na.rm=T),
                   MAD = mad(val, na.rm=T)) %>% 
  dplyr::mutate_if(is.numeric, round, digits=2) %>% 
  dplyr::mutate(Missing = paste0(Missing, "%"))
  
  
DT::datatable(infodf_summary, extensions = 'Buttons', options = list(dom = 'Bfrtip',buttons = c('csv')))





```

### Distributions of INFO Field Parameters
```{r plot_info_field, echo=F, include=T, warning=F, fig.width=11, fig.height=8}


plotObjAllInfo <- infodf %>% 
  tidyr::gather(key="INFO_field_param", val="val", 2:ncol(infodf)) %>%
  ggplot(aes(x=INFO_field_param, y=val)) + 
  geom_violin(aes(fill=factor(INFO_field_param))) +
  geom_boxplot(width=0.3) +
  geom_jitter(colour="#bdbdbd", height = 0.01, width = 0.01, size=0.2, alpha=0.1) +
  facet_wrap(~INFO_field_param, scales = "free") +
  ggtitle("Distribution of INFO Field Parameters") + 
  xlab("INFO Field Parameters") + ylab("Values") + 
  theme(panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(), 
        panel.background = element_blank(),
        panel.border = element_rect(color = "black", fill = NA, size = 0.5), 
        plot.title =  element_text(size=14, family = "Arial", face="bold", hjust=0.5, vjust=0.5),
        axis.ticks = element_blank(),
        axis.text.y = element_text(size=9, family = "Arial",  face="bold"),
        axis.title.y = element_text(size=13, family = "Arial", face="bold"),
        axis.text.x = element_blank(),
        axis.title.x = element_text(size=13, family = "Arial", face="bold"),
        strip.text.x = element_text(size =12, face="bold", family = "Arial"),
        legend.position = "none"
  )


plot(plotObjAllInfo)


```

### Distributions of GQ Format Field Parameter
Overall likelihood that the GT call that was made was correct. 

```{r plot_info_field, echo=F, include=T, warning=F, fig.width=11, fig.height=8}

formatdf <- tidyvcf$gt
formatdf %>% 
  select(ChromKey, Indiv, gt_GQ) %>% 
  ggplot(aes(x=ChromKey, y=gt_GQ)) + 
  geom_violin(aes(fill=factor(Indiv))) +
  geom_boxplot(width=0.3) +
  geom_jitter(colour="#bdbdbd", height = 0.01, width = 0.01, size=0.2, alpha=0.1) +
  facet_wrap(~Indiv, scales = "free") +
  ggtitle("Distribution of GQ Format Parameter") + 
  xlab("Sample") + ylab("GQ Values") + 
  theme(panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(), 
        panel.background = element_blank(),
        panel.border = element_rect(color = "black", fill = NA, size = 0.5), 
        plot.title =  element_text(size=14, family = "Arial", face="bold", hjust=0.5, vjust=0.5),
        axis.ticks = element_blank(),
        axis.text.y = element_text(size=9, family = "Arial",  face="bold"),
        axis.title.y = element_text(size=13, family = "Arial", face="bold"),
        axis.text.x = element_blank(),
        axis.title.x = element_text(size=13, family = "Arial", face="bold"),
        strip.text.x = element_text(size =12, face="bold", family = "Arial"),
        legend.position = "none"
  )


```
## Initial Impressions
words

### Population Allele Frequencies
words

### Allelic Depth
words

### Mapping Quality and Quality Depth
words

### Strand Odds Ratio
words

### Genotype Likelihood
words

## Apply Filters on Info Column
Applying these filters in R and will revisit VCF. 
```{r, echo=F, include=T, warning=F, fig.width=11, fig.height=8}
filtered_vcf <- NFBtools::vcffilter_info(vcfRobject = NULL,
                           infoMQ=55,
                           infoQD=25,
                           infoSOR=2,
                           infoAF = 0.05,
                           infoDP = NULL, # this is a percentile cutoff
                           infoFS = NULL,
                           infoMQRankSum = NULL,
                           infoReadPosRankSum = NULL,
                           biallelic = TRUE,
                           SNPs = FALSE)
```

## Apply Filters on Format Column
Applying these filters in R and will revisit VCF. 
```{r, echo=F, include=T, warning=F, fig.width=11, fig.height=8}
filtered_vcf <- NFBtools::vcffilter_format(vcfRobject = NULL,
                             formatGQ=NULL,
                             formatDP = NULL,
                             formatSP = NULL,
                             prop.loci.missing = 0.5, # this is given a loci, how many samples can have missing information before it is dropped
                             biallelic = TRUE,
                             SNPs = FALSE)
```

## Apply Filters on Segregating Sties
Applying these filters in R and will revisit VCF. 
```{r, echo=F, include=T, warning=F, fig.width=11, fig.height=8}
filtered_vcf <- NFBtools::vcfR2segsites(vcfRobj = NULL, err = 0.025)
```
_**This is the post-filtering VCF parameters:**_
filtered_vcf



## IGV Exploration
words










